ts/features/messages/saga/handleDownloadAttachment.ts

```diff
@@ -0,0 +1,181 @@
+import { getError } from "../../../utils/errors";
+import {
+  cancelPreviousAttachmentDownload,
+  downloadAttachment
+} from "../../../store/actions/messages";
+import {
+  UIAttachment,
+  UIMessageId
+} from "../../../store/reducers/entities/messages/types";
+import { ServiceId } from "../../../../definitions/backend/ServiceId";
+import { getServiceByMessageId } from "../../../store/reducers/entities/messages/paginatedById";
+import {
+  trackThirdPartyMessageAttachmentBadFormat,
+  trackThirdPartyMessageAttachmentDownloadFailed,
+  trackThirdPartyMessageAttachmentUnavailable
+} from "../analytics";
+import { getHeaderByKey } from "../utils/strings";
+import { handleRequestInit } from "./handleRequestInit";
+
+export const AttachmentsDirectoryPath =
+  RNFS.CachesDirectoryPath + "/attachments";
+
+/**
+ * Builds the save path for the given attachment
+ * @param attachment
+ */
+const savePath = (attachment: UIAttachment) =>
+  AttachmentsDirectoryPath +
+  "/" +
+  attachment.messageId +
+  "/" +
+  attachment.id +
+  "/" +
+  attachment.displayName;
+
+const getDelay = (headers: Record<string, string>) =>
+  pipe(
+    getHeaderByKey(headers, "retry-after"),
+    NumberFromString.decode,
+    E.map(delay => delay * 1000),
+    E.getOrElse(() => 0)
+  );
+
+function trackFailureEvent(
+  skipMixpanelTrackingOnFailure: boolean,
+  httpStatusCode: number,
+  messageId: UIMessageId,
+  serviceId: ServiceId | undefined
+) {
+  if (skipMixpanelTrackingOnFailure) {
+    return;
+  }
+  if (httpStatusCode === 500) {
+    trackThirdPartyMessageAttachmentUnavailable(messageId, serviceId);
+  } else if (httpStatusCode === 415) {
+    trackThirdPartyMessageAttachmentBadFormat(messageId, serviceId);
+  } else if (httpStatusCode <= 200 || httpStatusCode >= 400) {
+    trackThirdPartyMessageAttachmentDownloadFailed(messageId, serviceId);
+  }
+}
+
+export function* downloadAttachmentWorker(
+  bearerToken: SessionToken,
+  action: ActionType<typeof downloadAttachment.request>
+): SagaIterator {
+  const { skipMixpanelTrackingOnFailure, ...attachment } = action.payload;
+  const messageId = attachment.messageId;
+  const serviceId = yield* select(getServiceByMessageId, messageId);
+
+  while (true) {
+    try {
+      const config = yield* call(ReactNativeBlobUtil.config, {
+        path: savePath(attachment),
+        timeout: fetchTimeout
+      });
+
+      const { method, attachmentFullUrl, headers } = yield* call(
+        handleRequestInit,
+        attachment,
+        bearerToken,
+        uuid()
+      );
+
+      const result = yield* call(
+        config.fetch,
+        method,
+        attachmentFullUrl,
+        headers
+      );
+
+      const { status, ...rest } = result.info();
+
+      if (status === 200) {
+        const path = result.path();
+        yield* put(downloadAttachment.success({ attachment, path }));
+      } else if (status === 503) {
+        const waitingMs = getDelay(rest.headers);
+        if (waitingMs > 0) {
```

Comment:
If the header does not include a retry-after, we should treat the 503 as a general Service Unavailable and fail. That's why I would suggest that this condition should be
```
if (waitingMS >= 0) {
  yield* delay(waitingMs);
  continue;
}
throw Error(/* something catchable later */)
```

Reply:
Changed  https://github.com/pagopa/io-app/pull/5149/commits/0c149646134f9637673f6a982ce864adf95585e4